keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,1114 @@
+metadata:
+  name: pod1
+  namespace: default
+spec:
+[...]
+  networks:
+  - podNetwork: default
+    interfaceName: eth0
+    isDefaultGW4: true
+  - podNetwork: dataplane1
+    interfaceName: net1
+  - podNetworkAttachmentName: my-interface-fast
+    interfaceName: net2
+status:
+[...]
+  podIP: 192.168.5.54
+  podIPs:
+  - ip: 192.168.5.54
+    podNetwork: default
+    interfaceName: eth0
+  - ip: 10.0.0.20
+    podNetwork: dataplane1
+    interfaceName: net1
+  - ip: 2011::233
+    podNetwork: my-interface-fast
+    interfaceName: net2
+```
+The above status is expected to be populated by kubelet, but this can only happen
+after CRI provides support for the new Pod API. Because of that, initially
+kubelet will behave as it does today, without updating the additional fields.
+Until CRI catches up, the PodNetwork providers will be able to update that field
+on their own.
+
+#### DRA integration (alternative)
+We have discussed, as an alternative model, usage of the DRA API, and concluded
+that using it will be less clear for the user, compared to the explicit
+PodNetwork model, proposed above.
+
+### API server changes
+These are the changes for the API server covered by this design:
+* Provide validation webhook for PodNetwork or PodNetworkAttachment objects
+* Handle “default” PodNetwork deletion
+* Extend static Pod spec validation
+
+### Scheduler changes
+These are the changes we will do in Pod scheduler:
+* Provide active Pod spec validation
+
+When one of the multi-network validation fails, scheduler will follow the current
+“failure” path for Pod:
+* set PodScheduled condition (of the Pod) to False with appropriate error message
+* send Pod Event with same error message
+
+### Endpointslice controller changes
+Considering changes to Pod.PodStatus.PodIPs list, we must ensure that the
+controller is using the correct IPs when creating the endpoints. We will ensure
+that only IPs for "defautl" PodNetwork will be used.
+
+### Kubelet changes
+We will introduce an additional check for kubelet readiness for networking.
+Today kubelet does this verification via CRI that checks CNI config presence.
+We will add a “default” PodNetwork presence check to this flow. Until such a
+PodNetwork is not created, kubelet will keep the Ready condition in “False”
+with “default PodNetwork not found in API” error message.
+
+### CRI changes
+Considering the main input argument for kubelet when it interacts with CRI is
+the v1.Pod object, the above changes cover the kubelet-side part of providing
+the required data for multi-network. Next what is required are the changes to
+CRI API and CNI API which include
+* Pod creation flow update in RunPodSandbox
+* Pod status flow update in PodSandboxStatus
+* CNI input and output values
+
+Considering all above changes are in the direct scope of CRI, this KEP will not
+propose complete changes for them, and a separate KEP will be created to cover it.
+Below are some suggestions on what the changes could look like.
```

Comment:
Yes, I would see that we introduce this API as first step. Then container runtimes catch up and adjust to the new fields Pod has. Otherwise the change just grows too large. Lastly I can solve this with agent-based CNI Today, that does not rely on CRI, until they catch up.

Reply:
> Yes, I would see that we introduce this API as first step. Then container runtimes catch up and adjust to the new fields Pod has.

I think you need to provide more details on this, how much do you estimate container runtime catch up on features?

Reply:
IDK, considering KNI proposal is on the table as well, maybe then it will be independent. My point is, I do not want to make dependency on container runtimes.